chain/testing.go

```diff
@@ -175,9 +175,8 @@ func (f *Builder) Build(parent types.TipSet, width int, build func(b *BlockBuild
 		}
 
 		// Compute state root for this block.
-		tipSet, err := types.NewTipSet(b)
-		require.NoError(f.t, err)
-		b.StateRoot = f.ComputeState(tipSet)
+		prevState := f.StateForKey(parent.Key())
+		b.StateRoot, err = f.stateBuilder.ComputeState(prevState, [][]*types.SignedMessage{b.Messages})
```

Comment:
Previously the block was placed in a tipset before the state root was calculated, which computed and cached the wrong CID in the block 😬. That took me quite a while to track down - the cached CID is a bit of a trap!

Reply:
I believe this was the purpose of the `paranoid` field in the block -- the comment above it hints at a similar strategy to the one @ZenGround0 has proposed: https://github.com/filecoin-project/go-filecoin/blob/44b95abe40a20441561516b797dd3e46ab0f3939/types/block.go#L63-L66